src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,161 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        linkTail(value);
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == 0) {
+            linkHead(value);
+        } else if (index == size) {
+            linkTail(value);
+        } else {
+            Node<T> currentNode = findNode(index);
+            Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+            currentNode.prev.next = newNode;
+            currentNode.prev = newNode;
+            size++;
+        }
+
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            linkTail(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        Node<T> finderNode = findNode(index);
+        return finderNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> newNode = new Node<>(null, value, null);
+        T returnValue = get(index);
+        if (index == 0) {
+            head.next.prev = newNode;
+            newNode.next = head.next;
+            head = newNode;
+        } else {
+            Node<T> finderNode = findNode(index);
+            finderNode.next.prev = newNode;
+            finderNode.prev.next = newNode;
+        }
+
+        return returnValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T returnValue = get(index);
+        Node<T> deleteNode = findNode(index);
+        unlink(deleteNode);
+        return returnValue;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        boolean elementFound = false;
+        Node<T> deleteNode = head;
+        for (int i = 0; i < size; i++) {
+            if (deleteNode.value == null && object != null) {
+                deleteNode = deleteNode.next;
+            } else if ((deleteNode.value == null)
+                    || (deleteNode.value.equals(object))) {
+                unlink(deleteNode);
+                elementFound = true;
+                break;
+            } else {
+                deleteNode = deleteNode.next;
+            }
+        }
+        return elementFound;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> findNode(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Exception in thread \"main\" "
+                    + "java.lang.IndexOutOfBoundsException");
+        }
```

Comment:
extract that to the separate method (SRP violation)
```suggestion
        checkIndex(index);
```